root_numpy/_utils.py

```diff
@@ -145,6 +150,10 @@ def stretch(arr, fields=None):
             # Scalar field
             ret[field] = np.repeat(arr[field], len_array)
 
+    if return_indices:
+        idx = np.concatenate(list(map(np.arange, len_array)))
```

Comment:
I can't think of something very efficient than can be done without dropping into Cython.
Something that seems to be a bit faster if the array lengths are small is
```python
idx = np.fromiter(itertools.chain.from_iterable(map(range, len_array)))
```

Example:
```python
#!/usr/bin/env python
import numpy as np
import itertools

N = np.full(1000000, 10, dtype='int64')

def list_version(N):
    return np.concatenate(list(map(np.arange, N)))

def iter_version(N):
    return np.fromiter(itertools.chain.from_iterable(map(range, N)), dtype='int64')
```
And timing the function calls:
```python
In [6]: N = np.full(1000000, 100, dtype='int64')
In [7]: %timeit list_version(N)
1 loops, best of 3: 1.53 s per loop
In [8]: %timeit iter_version(N)
1 loops, best of 3: 3.93 s per loop
In [9]: N = np.full(1000000, 10, dtype='int64')
In [10]: %timeit list_version(N)
1 loops, best of 3: 1.26 s per loop
In [11]: %timeit iter_version(N)
1 loops, best of 3: 670 ms per loop
```

I assume that the iterator version also uses less memory.